shell/src/main/java/alluxio/cli/fs/command/FreeCommand.java

```diff
@@ -78,11 +79,16 @@ protected void runPlainPath(AlluxioURI path, CommandLine cl)
     try {
       CommonUtils.waitFor("file to be freed. Another user may be loading it.", () -> {
         try {
-          boolean freed = mFileSystem.getStatus(path).getInAlluxioPercentage() == 0;
-          if (!freed) {
+          URIStatus fileStatus = mFileSystem.getStatus(path);
+          if (fileStatus.getLength() == 0 && !fileStatus.isFolder()) {
+            // `getInAlluxioPercentage()` will always return 100,
+            // but 'free' on an empty file should be a no-op
+            return true;
+          }
+          if (fileStatus.getInAlluxioPercentage() >= 0) {
             mFileSystem.free(path, options);
```

Comment:
What will happen if `runPlainPath` is called for a folder with immediate 0byte files?

Reply:
Good point, let me look into/test that

Reply:
```czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs ls /
-rw-r--r--  czhu           czhu                         0 TO_BE_PERSISTED 05-03-2021 15:16:39:278 100% /emptyfile
drwxr-xr-x  czhu           czhu                         1       PERSISTED 05-03-2021 15:18:07:259  DIR /testdir
-rw-r--r--  czhu           czhu                     27040       PERSISTED 05-03-2021 15:17:06:958 100% /LICENSE
drwxr-xr-x  czhu           czhu                         1       PERSISTED 05-03-2021 15:18:28:568  DIR /emptyfiledir
drwxr-xr-x  czhu           czhu                         0   NOT_PERSISTED 05-03-2021 15:17:46:580  DIR /emptydir
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs free /emptyfile
/emptyfile was successfully freed from Alluxio space.
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs ls /
-rw-r--r--  czhu           czhu                         0       PERSISTED 05-03-2021 15:16:39:278 100% /emptyfile
drwxr-xr-x  czhu           czhu                         1       PERSISTED 05-03-2021 15:18:07:259  DIR /testdir
-rw-r--r--  czhu           czhu                     27040       PERSISTED 05-03-2021 15:17:06:958 100% /LICENSE
drwxr-xr-x  czhu           czhu                         1       PERSISTED 05-03-2021 15:18:28:568  DIR /emptyfiledir
drwxr-xr-x  czhu           czhu                         0   NOT_PERSISTED 05-03-2021 15:17:46:580  DIR /emptydir
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs free /LICENSE
/LICENSE was successfully freed from Alluxio space.
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs ls /
-rw-r--r--  czhu           czhu                         0       PERSISTED 05-03-2021 15:16:39:278 100% /emptyfile
drwxr-xr-x  czhu           czhu                         1       PERSISTED 05-03-2021 15:18:07:259  DIR /testdir
-rw-r--r--  czhu           czhu                     27040       PERSISTED 05-03-2021 15:17:06:958   0% /LICENSE
drwxr-xr-x  czhu           czhu                         1       PERSISTED 05-03-2021 15:18:28:568  DIR /emptyfiledir
drwxr-xr-x  czhu           czhu                         0   NOT_PERSISTED 05-03-2021 15:17:46:580  DIR /emptydir
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs free /emptydir
/emptydir was successfully freed from Alluxio space.
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs free /testdir
/testdir was successfully freed from Alluxio space.
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs free /emptyfiledir
/emptyfiledir was successfully freed from Alluxio space.
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs ls /emptydir
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs ls /testdir
-rw-r--r--  czhu           czhu                     27040       PERSISTED 05-03-2021 15:18:07:663   0% /testdir/LICENSE
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs ls /emptyfiledir
-rw-r--r--  czhu           czhu                         0       PERSISTED 05-03-2021 15:18:28:709 100% /emptyfiledir/emptyfile```

Reply:
```czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs cp /LICENSE /emptyfiledir
Copied /LICENSE to /emptyfiledir/LICENSE
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs ls /emptyfiledir
-rw-r--r--  czhu           czhu                         0       PERSISTED 05-03-2021 15:18:28:709 100% /emptyfiledir/emptyfile
-rw-r--r--  czhu           czhu                     27040       PERSISTED 05-03-2021 15:25:36:802 100% /emptyfiledir/LICENSE
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs free /emptyfiledir
/emptyfiledir was successfully freed from Alluxio space.
czhu@czhu-deb-laptop:~/src/github/personal/alluxio$ ./bin/alluxio fs ls /emptyfiledir
-rw-r--r--  czhu           czhu                         0       PERSISTED 05-03-2021 15:18:28:709 100% /emptyfiledir/emptyfile
-rw-r--r--  czhu           czhu                     27040       PERSISTED 05-03-2021 15:25:36:802   0% /emptyfiledir/LICENSE```

Reply:
This behaviour for empty files inside of directories should be identical to before this PR; this PR only changes the behaviour of `FreeCommand` on the top-level inode.

If you look at [DefaultFileSystemMaster#freeInternal()](https://github.com/Alluxio/alluxio/blob/40af661d15104813f74eeda3790b7a366f7cb28f/core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java#L2603) it will just process the empty file "normally" (it will just have no blocks AFAIK).

Reply:
Makes perfect sense. Thanks for comprehensive testing.